### ISSUE TITLE
Allow generic adaptations to support layer names

### ISSUE THREAD

**Issue by christofmuc** [christofmuc/KnobKraft-orm#155]
The layer name capability is missing for the generic adaptation is only available in C++. I use it with the Rev2, and that is C++ code where a lot of more thing were possible. I'll create a new ticket for that. Might also make sense for the Pro2 and Pro3, I think those are layered as well.

_Originally posted by @christofmuc in https://github.com/christofmuc/KnobKraft-orm/issues/127#issuecomment-1108314776_

I could allow the layer name capability to be part of the Generic Adaptation in Python as well, just two functions to be implemented.

**Comment by Andy2No** [comment]
Cool.  Maybe you could just keep nameFromDump() as it is, and add a function for an optional second layer name, which is allowed to return null, or an empty string.

I still have the python adaptation for the Prophet 12, which looks fairly simple... unless that secretly relies on some C++ that I can't see, for part of what it does.   If that adaptation still works, it should be fairly simple to extend it for the new function.

It would be nice if it could be done without having to call unescapeSysex() twice - once for each name.  Since it loops over the whole patch, it must be fairly time consuming.

I don't know enough about python to know the best way to handle that - maybe a local copy of the unescaped sysex, and a flag to say it's been filled in, but that would have to only persist for the current patch.

Maybe one way would be to require the two name functiosn to be called in order, so nameFromDump() stores the unescaped form of the sysex in a shared variable, and the second one, e.g. nameFromDump2ndLayer(), then uses that data, rather than generating it again.

**Comment by christofmuc** [comment]
@Andy2No Thinking about this again, I think we shouldn't store local data in the Adaptation (you could by using global variables i the script and the global keyword to allow a method to modify that global). Because if you skip through tons of data, it will start piling up very quickly. 

Layer names could (should) be cached by the main program and calculated only once during load of the patch, unless they are changed. I think this is how it is done for single patch names, but layer names would need kind of an extension of that mechanism. Anyway, still don't optimize what you haven't measured, so I'd stick with the naive implementation first and then see if we have a problem...

**Comment by Andy2No** [comment]
I've been thinking another way to do it would be to calculate which 8 byte blocks the name or names fall into, and just unescape those, rather than the whole patch.  Since we know how to address the names in the unescaped output, we could translate that to addresses in the encoded data (starting a few bytes earlier, where needed, on an 8 byte boundary), then just process the encoded blocks that contain those bytes, rather than the whole patch, twice.

It's another opportunity to get things wrong though, so maybe it's not worth the effort.

**Comment by christofmuc** [comment]
@Andy2No This is released with today's 1.16.0. I did it in the generic way, and now the generic sequential adaptations allows to configure it (I think Pro 2 and Pro 3 also have these layers). Performance I didn't do anything special, so let's see how it behaves. It is a bit slower on my box than before, but still very ok.

**Comment by Andy2No** [comment]
@christofmuc Cool.  I'll look into whether the same approach applies to showing the B layer on the Tetra, at some point.  It's another member of the Prophet 08 family, between the Mopho and the Prophet 08 and Rev 2, so there's a lot in common with them all, apart from the combo (combi?) mode on the Tetra, which is unique, I think.

I've imported some saved patches from the Prophet 12, and those work fine... I notice long layer names get word wrapped, which takes up extra vertical space, but I can see why that might be tricky to fix, now the width of the box is adjustable - unless there's a feature of JUCE you can use to turn that off.

I've very recently taken delivery of a 16 voice Rev 2 Desktop, so I've got that set up at the moment, and I can only really have one synth set up at once, easily, so I'm sticking with that one for at least the next few days.

**Comment by christofmuc** [comment]
@Andy2No I think many DSIs have the layering - Pro 2, Pro 3, Prophet 12, Tetra, Rev 2... It is very easy to add in for all that are based on the Generic Sequential adaptation, we just need to revisit all the manuals.

And remember there is now the switch what to display on the button - you can show the layer names (default), or select to see just the number or only the patch name (aka layer A).

**Comment by Andy2No** [comment]
Yes, I like that option.  I noticed it doesn't change anything if I'm looking at All patches, though.  I realise that's a slightly different view of the patches, because it shows the synth name at the bottom of each one... Maybe the menu should just be hidden, for "All patches"?

It's not important, now I know that's what happens, but it baffled me for a couple of minutes.
